src/main/java/org/mockito/internal/util/KotlinInlineClassUtil.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockito.internal.util;
+
+import org.mockito.internal.stubbing.BaseStubbing;
+import org.mockito.internal.stubbing.OngoingStubbingImpl;
+import org.mockito.invocation.MatchableInvocation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class KotlinInlineClassUtil {
+    // When mocking function, returning inline class, its return type is
+    // underlying type.
+    // So, `thenReturn` calls fails, because of non-compatible types.
+    public static boolean isInlineClassWithAssignableUnderlyingType(
+            Class<?> inlineClass, Class<?> underlyingType) {
+        try {
+            // All inline classes have 'box-impl' method, which accepts
+            // underlying type and returns inline class.
+            inlineClass.getDeclaredMethod("box-impl", underlyingType);
+            return true;
+        } catch (NoSuchMethodException e) {
+            return false;
+        }
+    }
+
+    private static Object unboxInlineClass(Object boxedValue)
+            throws InvocationTargetException, IllegalAccessException {
+        Class<?> inlineClass = boxedValue.getClass();
+        Method unboxImpl = null;
+        for (Method declaredMethod : inlineClass.getDeclaredMethods()) {
+            if (declaredMethod.getName().equals("unbox-impl")) {
+                unboxImpl = declaredMethod;
+                break;
+            }
+        }
+
+        assert unboxImpl != null : inlineClass.getName() + " is not a Kotlin inline class";
+
+        return unboxImpl.invoke(boxedValue);
+    }
+
+    private static Class<?> getReturnTypeOfInvocation(BaseStubbing<?> stubbing) {
+        if (!(stubbing instanceof OngoingStubbingImpl<?>)) {
+            return null;
+        }
+        // TODO: Hack to get return type of method
+        OngoingStubbingImpl<?> ongoingStubbing = (OngoingStubbingImpl<?>) stubbing;
+        if (ongoingStubbing.getRegisteredInvocations().size() == 0) return null;
+
+        MatchableInvocation invocationForStubbing = ongoingStubbing.getInvocationForStubbing();
```

Comment:
I think we can use `org.mockito.internal.stubbing.answers.InvocationInfo` here, in a similar fashion to how `Returns` uses it: https://github.com/mockito/mockito/blob/80374488162380d006b8ae6da915d940cbf68c7c/src/main/java/org/mockito/internal/stubbing/answers/Returns.java#L32

In that case, we can augment `Returns.answer` to use the `InvocationInfo` API and we don't need to update the `OngoingStubbingImpl` interface. E.g. we move `unboxUnderlyingValueIfNeeded` into `Returns.answer`. We can then lazily compute `value` when it is invoked, to be minimally computational impactful.